src/runtime_config/default/actions.js

```diff
@@ -197,6 +197,84 @@ function setNodeCertificateValidityPeriod(
   );
 }
 
+function checkRecoveryThreshold(config, new_config) {
+  const from = config.recovery_threshold;
+  const to = new_config.recovery_threshold;
+  if (to === undefined || from === to) {
+    return;
+  }
+
+  const service_info = "public:ccf.gov.service.info";
+  const rawService = ccf.kv[service_info].get(getSingletonKvKey());
+  if (rawService === undefined) {
+    throw new Error("Service information could not be found");
+  }
+
+  const service = ccf.bufToJsonCompatible(rawService);
+
+  if (service.status === "WaitingForRecoveryShares") {
+    throw new Error(
+      `Cannot set recovery threshold if service is ${service.status}`
+    );
+  } else if (service.status === "Open") {
+    let activeRecoveryMembersCount = getActiveRecoveryMembersCount();
+    if (new_config.recovery_threshold > activeRecoveryMembersCount) {
+      throw new Error(
+        `Cannot set recovery threshold to ${new_config.recovery_threshold}: recovery threshold would be greater than the number of recovery members ${activeRecoveryMembersCount}`
+      );
+    }
+  }
+}
+
+function checkReconfigurationType(config, new_config) {
+  const from = config.reconfiguration_type;
+  const to = new_config.reconfiguration_type;
+  if (from !== to && to !== undefined) {
+    if (!(from === "OneTransaction" && to === "TwoTransaction")) {
+      throw new Error(
+        `Cannot change reconfiguration type from ${from} to ${to}.`
+      );
+    }
+  }
+}
+
+function updateServiceConfig(new_config) {
+  const service_config_table = "public:ccf.gov.service.config";
+  const rawConfig = ccf.kv[service_config_table].get(getSingletonKvKey());
+  if (rawConfig === undefined) {
+    throw new Error("Service configuration could not be found");
+  }
+  let config = ccf.bufToJsonCompatible(rawConfig);
+
+  // First run all checks
+  checkReconfigurationType(config, new_config);
+  checkRecoveryThreshold(config, new_config);
+
+  // Then all updates
+  if (new_config.reconfiguration_type !== undefined) {
+    config.reconfiguration_type = new_config.reconfiguration_type;
+  }
+
+  let need_recovery_threshold_refresh = false;
```

Comment:
Do we need this intermediate variable at all or could we call `ccf.node.triggerRecoverySharesRefresh();` directly on line `264` instead? `triggerRecoverySharesRefresh()` has no side effect apart from writing to the KV.

Reply:
I think we could, but it used to be done after the `.set`. If something goes wrong there, it would be called without the new value being recorded. I'm not sure whether something _can_ go wrong there though.

Reply:
Because of https://github.com/microsoft/CCF/issues/1031, we used to tend to write proposal actions as if the proposal execution wasn't transactional. This is no longer the case and all transaction execution is now transactional so it doesn't really matter which order you call operations on the KV in.

Reply:
Great, then I'll simplify this!